lib/SILOptimizer/SILCombiner/SILCombinerMiscVisitors.cpp

```diff
@@ -781,75 +782,137 @@ SILCombiner::visitInjectEnumAddrInst(InjectEnumAddrInst *IEAI) {
-    // We can localize the store to an alloc_stack.
-    // Allowing us to perform the same optimization as for the store.
-    //
-    //  %alloca = alloc_stack
-    //            apply(%alloca,...)
-    //  %load = load %alloca
-    //  %1 = enum $EnumType, $EnumType.case, %load
-    //  store %1 to %nopayload_addr
-    //
-    if ((AI = dyn_cast<ApplyInst>(&*II))) {
-      unsigned ArgIdx = 0;
-      for (auto &Opd : AI->getArgumentOperands()) {
-        // Found an apply that initializes the enum. We can optimize this by
-        // localizing the initialization to an alloc_stack and loading from it.
-        DataAddrInst = dyn_cast<InitEnumDataAddrInst>(Opd.get());
-        if (DataAddrInst && DataAddrInst->getOperand() == IEAI->getOperand() &&
-            ArgIdx < AI->getSubstCalleeType()->getNumIndirectResults()) {
-          EnumInitOperand = &Opd;
-          break;
-        }
-        ++ArgIdx;
-      }
-      // We found an enum initialization.
-      if (EnumInitOperand)
-        break;
-      AI = nullptr;
+    if (auto *DataAddrInst = dyn_cast<InitEnumDataAddrInst>(currUser)) {
+      DataAddrVec.push_back(DataAddrInst);
+      continue;
+    }
+    if (auto *AddrInst = dyn_cast<InjectEnumAddrInst>(currUser)) {
+      InjectEnumAddrVec.push_back(AddrInst);
+      continue;
+    }
+    if (currUser->mayWriteToMemory()) {
+      MayWriteSet.insert(currUser);
     }
   }
-  // Found the store to this enum payload. Check if the store is the only use.
+
+  if (DataAddrVec.size() != 1 || InjectEnumAddrVec.size() != 1) {
+    return nullptr;
+  }
+  assert((InjectEnumAddrVec.front() == IEAI) &&
+         "Found InitEnumDataAddrInst differs from IEAI");
+  InitEnumDataAddrInst *DataAddrInst = DataAddrVec.front();
+  // Found the DataAddrInst to this enum payload. Check if it has only use.
   if (!DataAddrInst->hasOneUse())
     return nullptr;
 
+  StoreInst *SI = dyn_cast<StoreInst>(DataAddrInst->use_begin()->getUser());
+  ApplyInst *AI = dyn_cast<ApplyInst>(DataAddrInst->use_begin()->getUser());
+  if (!SI && !AI) {
+    return nullptr;
+  }
+
+  // Make sure the enum pattern instructions are the only ones which write to
+  // this location
+  if (!MayWriteSet.empty()) {
+    // Analyze the instructions (implicit dominator analysis)
+    // If we find any of MayWriteSet, return nullptr
+    SILBasicBlock *InitEnumBB = DataAddrInst->getParent();
+    SILBasicBlock *InjectEnumBB = IEAI->getParent();
+    assert(InitEnumBB && "DataAddrInst is not in a valid Basic Block");
+    assert(InjectEnumBB && "IEAI is not in a valid Basic Block");
+    llvm::SmallVector<SILBasicBlock *, 64> Worklist;
+    Worklist.push_back(InjectEnumBB);
+    llvm::SmallSetVector<SILBasicBlock *, 16> Preds;
+    Preds.insert(InjectEnumBB);
+    bool ReachedInitEnumBB = false;
+    while (!Worklist.empty()) {
+      SILBasicBlock *BB = Worklist.pop_back_val();
+
+      // Check current BB for any of the MayWriteSet instructions
+      for (auto &InsIt : *BB) {
```

Comment:
Actually I am re-writing this code, iterating over the may write set is not enough. @eeckstein correctly pointed out that other memory-access instructions might be there. and that they might alias to the same address.